### PR TITLE
ENGINFRA-727: Added exception traceback decorator

### PR DESCRIPTION
**Pull request by hsaikia**
Exception stacktraces can become long and difficult to parse through in case of runtime errors. This PR introduces a decorator `exception_traceback` that wraps any function to return the expected value in case of success or prints a size limited exception traceback. The size of the list is configurable.

Edit : Added an additional `reraise` flag to re-reraise the exception in cases when we do not want the exception to be swallowed. This is of course equivalent to not using the decorator at all. However, in a follow up PR @jayatisarkar brings in more improvements, such as a list if exceptions that are allowed to be swallowed.

Note : This PR got messed up with some weird rebasing that went wrong. So I'll close this one. I opened a new PR here : https://github.com/kraken-build/kraken-core/pull/40

### CODE REVIEW

File: src/kraken/core/formatting/exception_traceback.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            try:
+                return func(*args, **kwargs)
+            except Exception:
+                if stacktrace_depth < 1:
+                    return None
+                traceback_lines = traceback.format_exc().splitlines()
+                return traceback_lines[-stacktrace_depth:]
```
[the remark this answers is not in the document]
Comment (hsaikia) [reply src/kraken/core/formatting/exception_traceback.py:17]: Got it. I implemented it this way in order to be able to test the output easily, as well as leave it to the caller to format the strings however they please. In my initial implementation the return type was still `T | None` and the exception stacktrace was just being printed. I guess I could revert back to that implementation.

[the remark this answers is not in the document]
Comment (hsaikia) [reply src/kraken/core/formatting/exception_traceback.py]: Fixed

File: src/kraken/core/formatting/exception_traceback.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import traceback
+import functools
+
+
+def exception_traceback(stacktrace_depth=2):
```
Comment (jayatisarkar) [inline src/kraken/core/formatting/exception_traceback.py:5]: I think you need add the types for the arguments in each of these methods

**Comment by samrogerson** [comment]
It looks like this wrapper swallows the exception, which is likely unintentional?

File: src/kraken/core/formatting/exception_traceback.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    def exception_handler(func):
+        """Decorator that formats an exception and returns only a part of the stacktrace as specified by the user."""
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            try:
+                return func(*args, **kwargs)
+            except Exception:
```
Comment (samrogerson) [inline src/kraken/core/formatting/exception_traceback.py:13]: It's probably worth re-raising the exception to prevent swallowing it entirely.

Unfortunately, doing so will mean the exception will reach the interpreter which will generate a full printout ;)

Options are to (a) alter the exception before re-raising it (which feels janky) (b) look at an alternative implementation using something like [`sys.excepthook`](https://docs.python.org/3/library/sys.html#sys.excepthook)

Comment (hsaikia) [reply src/kraken/core/formatting/exception_traceback.py]: @samrogerson I tested an alternative where I assigned sys.excepthook to a formatter function, and raised the exception. This however was similar to the current implementation - where the Exception is raised if the re-raise flag is set. One might argue that this flag is unnecessary, as not using the decorator essentially has the same behaviour. However, the flag makes it more explicit for the user, and with some added fixes on top of this PR coming up (https://github.com/kraken-build/kraken-core/pull/38) I believe this should be fine.

File: src/kraken/core/formatting/exception_traceback.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs) -> Optional:
+            try:
+                return func(*args, **kwargs)
+            except Exception as e:
+                if reraise:
+                    raise e
+                else:
+                    format_exception(traceback.format_exc().splitlines()[-stacktrace_depth:])
```
Comment (NiklasRosenstein) [inline src/kraken/core/formatting/exception_traceback.py:22]: `traceback.format_exc()` has a `limit` parameter that you might want to use instead.

File: src/kraken/core/formatting/exception_traceback.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+            for line in tb_lines:
+                print(f"> {line}")
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs) -> Optional:
+            try:
+                return func(*args, **kwargs)
+            except Exception as e:
+                if reraise:
+                    raise e
```
[the remark this answers is not in the document]
Comment (hsaikia) [reply src/kraken/core/formatting/exception_traceback.py:20]: There is actually none (as I mentioned in the description) apart from making it more explicit. I can completely remove it as well if we can agree that using the decorator makes it explicit enough to the user that the exception will be swallowed.

File: src/kraken/core/formatting/exception_traceback.py (hunk 4 above)
Comment (hsaikia) [reply src/kraken/core/formatting/exception_traceback.py:22]: I did try it but it didn't seem to output the same result. Let me try again.

File: src/kraken/core/formatting/exception_traceback.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+from typing import Callable, Optional
+
+
+def exception_traceback(stacktrace_depth: int = 2, reraise: bool = True) -> Callable:
+    def exception_handler(func: Callable) -> Callable:
+        """Decorator that formats an exception and returns only a part of the stacktrace as specified by the user."""
+
+        def format_exception(tb_lines: list[str]):
+            for line in tb_lines:
+                print(f"> {line}")
```
[the remark this answers is not in the document]
Comment (hsaikia) [reply src/kraken/core/formatting/exception_traceback.py]: Fixed

File: src/kraken/core/formatting/exception_traceback.py (hunk 4 above)
Comment (hsaikia) [reply src/kraken/core/formatting/exception_traceback.py]: Done.

File: src/kraken/core/formatting/exception_traceback.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import sys
+import traceback
+import functools
+from typing import Callable, Optional
+
+
+def exception_traceback(stacktrace_depth: int = 2, reraise: bool = True) -> Callable:
```
Comment (NiklasRosenstein) [inline src/kraken/core/formatting/exception_traceback.py:7]: Can you add to this PR a practical use of this decorator? I.e. using the decorator in the code base where you thought it should be used.

File: src/kraken/core/formatting/exception_traceback.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hsaikia) [reply src/kraken/core/formatting/exception_traceback.py:21]: Right. I tested with both formatting on and off, and saw no real difference in case when we re-raise - i.e., there was no change to the visual clutter and the same information was being propagated. I could for sure add a formatted line in such a case.

File: src/kraken/core/formatting/exception_traceback.py (hunk 7 above)
Comment (hsaikia) [reply src/kraken/core/formatting/exception_traceback.py:7]: Sure. I can add the decorator to a few spots where I observed errors.

File: src/kraken/core/formatting/exception_traceback.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hsaikia) [reply src/kraken/core/formatting/exception_traceback.py:21]: Done.

File: src/kraken/core/formatting/exception_traceback.py (hunk 7 above)
Comment (hsaikia) [reply src/kraken/core/formatting/exception_traceback.py:7]: Done.
